test_conformance/extensions/cl_khr_command_buffer/negative_command_buffer_copy.cpp

```diff
@@ -0,0 +1,503 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "basic_command_buffer.h"
+#include "procs.h"
+#include <vector>
+
+//--------------------------------------------------------------------------
+namespace {
+
+// CL_INVALID_COMMAND_QUEUE if command_queue is not NULL.
+struct CommandBufferCopyQueueNotNull : public BasicCommandBufferTest
+{
+    using BasicCommandBufferTest::BasicCommandBufferTest;
+
+    cl_int Run() override
+    {
+        cl_int error =
+            clCommandCopyBufferKHR(command_buffer, queue, in_mem, out_mem, 0, 0,
+                                   data_size(), 0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_COMMAND_QUEUE,
+                               "clCommandCopyBufferKHR should return "
+                               "CL_INVALID_COMMAND_QUEUE",
+                               TEST_FAIL);
+
+        const size_t origin[3] = { 0, 0, 0 }, region[3] = { 512, 512, 1 };
+
+        error = clCommandCopyBufferRectKHR(
+            command_buffer, queue, in_mem, out_mem, origin, origin, region, 0,
+            0, 0, 0, 0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_COMMAND_QUEUE,
+                               "clCommandCopyBufferRectKHR should return "
+                               "CL_INVALID_COMMAND_QUEUE",
+                               TEST_FAIL);
+
+        const cl_image_format formats = { CL_RGBA, CL_UNSIGNED_INT8 };
+        clMemWrapper image = create_image_2d(
+            context, CL_MEM_READ_WRITE, &formats, 512, 512, 0, NULL, &error);
+        test_error(error, "create_image_2d failed");
+        clMemWrapper buffer = clCreateBuffer(context, CL_MEM_READ_WRITE,
+                                             data_size(), nullptr, &error);
+        test_error(error, "Unable to create buffer");
+
+        error = clCommandCopyImageToBufferKHR(command_buffer, queue, image,
+                                              buffer, origin, region, 0, 0,
+                                              nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_COMMAND_QUEUE,
+                               "clCommandCopyImageToBufferKHR should return "
+                               "CL_INVALID_COMMAND_QUEUE",
+                               TEST_FAIL);
+
+        return CL_SUCCESS;
+    }
+};
+
+// CL_INVALID_CONTEXT if the context associated with command_queue,
+// command_buffer, src_buffer, and dst_buffer are not the same.
+struct CommandBufferCopyDifferentContexts : public BasicCommandBufferTest
+{
+    CommandBufferCopyDifferentContexts(cl_device_id device, cl_context context,
+                                       cl_command_queue queue)
+        : BasicCommandBufferTest(device, context, queue), in_mem_ctx(nullptr),
```

Comment:
You can put all these nullptr initialisers as designated initialisers on the declaration line rather than list them on the constructor which will be more cumbersome.